### ISSUE TITLE
Declarations for function-scoped class expressions

### ISSUE THREAD

**Issue by awerlang** [Microsoft/TypeScript#3129]
TLDR: How to generated .d.ts from function-scoped class expressions?

Consider the following snippet:

```ts
class MyService {
	constructor(private dependency) { }
	myMethod() {
		return this.dependency.someMethod();
	}
}

function myFactory(dependency) {
	return new MyService(dependency);
}
```

As you can see, I have to make a field for a dependency. This is so C# (I like C#) but I'd prefer the JavaScript way in JavaScript.

Then I make a closure:
 
```ts
function myFactory2(dependency) {
	return new class MyService {
		myMethod() {
			return dependency.someMethod();
		}
	};
}
```

I'm aware that I get a new constructor function each time and there's no prototype sharing, but I don't care. I see a high potential in this pattern for true **information hiding**. It is even better for inheritance. A base class could evolve without breaking derived classes.

To make it fit on TypeScript, however, automatic generation of declarations or somehow compiler being able to interpret this pattern is needed I think. An identifier for this factory type is needed.

One should be able to do the following then, based on the second snippet:

```ts
function MyController(myService: myFactory2) {
}
```

**Comment by ahejlsberg** [comment]
This is slated for the 1.6 release: https://github.com/Microsoft/TypeScript/wiki/Roadmap.

**Comment by ahejlsberg** [comment]
Also see #497.
